### PR TITLE
#221: Added an empty manifest to script language containers

### LINKED ISSUE #221: Add an empty manifest to script language containers
##Summary
When uploading an Script Language Container to a database, there is currently no mechanism available to check if the Script-Language-Container is ready to use on all nodes. The distribution of the Script-Languages-Container includes several steps, like copying the TAR file to all nodes and decompressing it. The current workaround, which is to wait for n seconds, is a poor solution.

We want to add a file at the end of the TAR file, which can be used later by some UDF’s to check for. As TAR is a sequential file format, and decompression happens sequentially in COS, it’s safe to check only for this file. Once it’s available on all nodes, clients can be sure, that the uploaded Script-Languages-Container is ready-to-use.

### PR DESCRIPTION
**Pull request by tomuben**
fixes #221

### CODE REVIEW

File: exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py [hunk 1]
```diff
@@ -153,8 +153,20 @@ def _export_container(self, container, release_image_name: str, temp_directory:
     def _pack_release_file(self, log_path: Path, extract_dir: str, release_file: Path):
         self.logger.info("Pack container file %s", release_file)
         extract_content = " ".join("'%s'" % file for file in os.listdir(extract_dir))
```
Comment (ahsimb) [inline exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py:155]:
```suggestion
        extract_content = " ".join(f"'{file}'" for file in os.listdir(extract_dir))
```

**Review by kaklakariada — DISMISSED** [review]

File: exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py [hunk 2]
```diff
@@ -153,8 +153,20 @@ def _export_container(self, container, release_image_name: str, temp_directory:
...
-        self.run_command(command, "packing container file %s" % release_file,
+        if not str(release_file).endswith("tar.gz"):
+            raise ValueError(f"Unexpected release file: '{release_file}'. Expected suffix 'tar.gz'.")
+        tmp_release_file = release_file.with_suffix("") #cut off ".gz" from ".tar.gz"
+        command = f"""tar -C '{extract_dir}' -vcf '{tmp_release_file}' {extract_content}"""
+        self.run_command(command, "packing container file %s" % tmp_release_file,
+                         log_path.joinpath("pack_release_file.log"))
+        manifest_file = os.path.join(extract_dir, "exasol-manifest.json")
+        with open(manifest_file, "w") as f:
+            pass
```
[the remark this answers is not in the document]
Comment (tomuben) [reply exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py:164]: Hm, good point actually.

File: exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py [hunk 3]
```diff
@@ -153,8 +153,20 @@ def _export_container(self, container, release_image_name: str, temp_directory:
     def _pack_release_file(self, log_path: Path, extract_dir: str, release_file: Path):
         self.logger.info("Pack container file %s", release_file)
         extract_content = " ".join("'%s'" % file for file in os.listdir(extract_dir))
-        command = f"""tar -C '{extract_dir}' -cvzf '{release_file}' {extract_content}"""
-        self.run_command(command, "packing container file %s" % release_file,
+        if not str(release_file).endswith("tar.gz"):
+            raise ValueError(f"Unexpected release file: '{release_file}'. Expected suffix 'tar.gz'.")
+        tmp_release_file = release_file.with_suffix("") #cut off ".gz" from ".tar.gz"
+        command = f"""tar -C '{extract_dir}' -vcf '{tmp_release_file}' {extract_content}"""
+        self.run_command(command, "packing container file %s" % tmp_release_file,
```
[the remark this answers is not in the document]
Comment (tomuben) [reply exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py:160]: Fixed

File: exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py [hunk 4]
```diff
@@ -153,8 +153,20 @@ def _export_container(self, container, release_image_name: str, temp_directory:
...
+                         log_path.joinpath("pack_release_file.log"))
+        manifest_file = os.path.join(extract_dir, "exasol-manifest.json")
+        with open(manifest_file, "w") as f:
+            pass
+        command = f"""tar -C '{extract_dir}' -rvf '{tmp_release_file}' exasol-manifest.json"""
+        self.run_command(command, "adding manifest to %s" % tmp_release_file,
```
Comment (ahsimb) [inline exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py:166]:
```suggestion
        self.run_command(command, f"adding manifest to {tmp_release_file}",
```

File: exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py [hunk 5]
```diff
@@ -153,8 +153,20 @@ def _export_container(self, container, release_image_name: str, temp_directory:
...
+                         log_path.joinpath("pack_release_file.log"))
+        command = f"""gzip {tmp_release_file}"""
+        self.run_command(command, "Creating %s" % release_file,
```
Comment (ahsimb) [inline exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py:169]:
```suggestion
        self.run_command(command, f"Creating {release_file}",
```

File: exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py (hunk 1 above)
Comment (tomuben) [reply exasol_script_languages_container_tool/lib/tasks/export/export_container_base_task.py:155]: Fixed

**Review by ahsimb — APPROVED** [review]

**Pull request merged**